[ownership] update mem2reg to support ossa

zoecarver opened the pull request:
<!-- What's in this pull request? -->
This patch updates mem2reg to support ossa. I used destructure wherever possible like suggested in #30278. 

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

zoecarver committed:
Add ossa tests

zoecarver committed:
Remove ownership check in mem2reg

zoecarver committed:
Add support for begin/end access/borrow

zoecarver committed:
Add tuple destructure support

zoecarver commented on lib/SIL/SILOwnershipVerifier.cpp:
```diff
@@ -841,6 +841,7 @@ void SILInstruction::verifyOperandOwnership() const {
 
     assert(errorBehavior.shouldAssert() &&
            "At this point, we are expected to assert");
+    opValue->getFunction()->dump();
```
Oops I'll remove this.

zoecarver commented on test/SILOptimizer/mem2reg_ossa.sil:
```diff
@@ -0,0 +1,485 @@
+// RUN: %target-sil-opt -enable-sil-verify-all %s -mem2reg | %FileCheck %s
```
Again, it would probably be good to add some non-trivial tests.

gottesmm replied:
Can you add that.

gottesmm replied:
Also, if you think this is really ready for ossa, I wonder what the effect would be if we ran this before semantic arc opts? Before we do that, I am probably going to need to move the eager specializer though to get a nice effect (a bunch of allocations are needed due to in_guaranteed args.

zoecarver replied:
For sure. I wanted to make sure that this patch was going in the right direction before writing hundreds of lines of tests that would need to be re-written if I had to significantly change the direction of this patch.

zoecarver commented:
I found some issues while compiling the stdlib. I'll fix those now (I meant to run that check before submitting this patch, sorry).

zoecarver committed:
Fix tuple with half trivial types

zoecarver committed:
Fix bbargs

Check that there's a use of the stack_alloc in the bb that is having an
arg added to it.

If there's an arg added, make sure that the value of the arg is copied
(unless it's undef).

Also, formatting.

zoecarver committed:
Fix consumption after removal of store

zoecarver committed:
Move consume store src out of loop

zoecarver committed:
Remove begin_borrow (it doesn't make sense because its operand wont be an address) and bail on loads in multiple blocks

zoecarver committed:
Bail on load [take]

zoecarver committed:
Explicitly check for store source consuming users and bail on undefs

zoecarver committed:
Only create copy_value in ownership mode

zoecarver committed:
Update tests

zoecarver committed:
Fix some issues with load values being destroyed or not destroyed

zoecarver committed:
Drop support for load in arg promotion

zoecarver committed:
Fix gaurenteed ownership type use

zoecarver committed:
Fixed the following based on stdlib compiliation failures:

* Don't remove destroys unless there are no other users remaining
* Unless there are no consuming uses, copy before adding a branch
argument
* Bail on non-trivial ownership loads/stores in phi

zoecarver commented on lib/SILOptimizer/Transforms/SILMem2Reg.cpp:
```diff
@@ -547,23 +646,48 @@ void MemoryToRegisters::removeSingleBlockAllocation(AllocStackInst *ASI) {
         NumInstRemoved++;
-        // No need to continue scanning after deallocation.
-        break;
+        continue;
```
To catch `destroy_values`.

zoecarver commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -614,9 +614,9 @@ void swift::eraseUsesOfInstruction(SILInstruction *inst, CallbackTy callback) {
       // nothing uses this instruction.
       eraseUsesOfInstruction(user, callback);
 
+      callback(user);
```
This patch is the only time `callback` is actually used so, this change is OK.

zoecarver commented on test/SILOptimizer/mem2reg_ossa.sil:
```diff
@@ -0,0 +1,978 @@
+  return %55 : $()
+}
+
+// TODO:
```
I've got these all in the form of hundred line stdlib reproducers. Today (hopefully) I'll turn them into nice tests (like the ones above).

zoecarver committed:
Clean up and formatting

zoecarver commented on lib/SILOptimizer/Transforms/SILMem2Reg.cpp:
```diff
@@ -361,19 +380,59 @@ static void replaceLoad(LoadInst *LI, SILValue val, AllocStackInst *ASI) {
+        userOp->isConsumingUse() && !isa<DestroyValueInst>(userOp->getUser()))
+      userOp->set(builder.createCopyValue(LI->getLoc(), val));
+    // If the user can't accept the projection, create a copy/destroy and hope
+    // that fixes it. Currently, the only time that this would happen is if
```
Can you verify this is a correct statement:
> Currently, the only time that this would happen is if we're replacing a guaranteed object with an owned one.

zoecarver commented:
Few... I spent the better part of the week working on updating this pass. It looks like the stdlib now compiles successfully (and we're getting some pretty great performance improvements on ossa!).

While updating the pass I did notice a few things which I'm planning on making separate patches for:

**`SILUndef`** 
This patch uses `SILUndef` a lot. It seems to use that when the optimization it's trying doesn't work (instead of bailing). Because of this, I've added a lot of guards for `isa<SILUndef>`. I'd like to remove as many uses of `SILUndef` as possible and ideally remove it from this pass altogether. I'm worried that it will slip into actual code (because before the guards, that was happening).

**Copy/destroy value elimination**
I'm sure there's another pass that does this but it would probably make sense to have basic copy/destroy value elimination before/after this pass. Because there are multiple steps in this pass, the following happens semi-frequently:
```
// In
%2 = unchecked_addr_cast
%3 = load [copy] %2 : $*Klass

// Out
%2 = unchecked_bitwise_cast
%3 = copy_value
destroy_value %3
```
While I don't think it's a big issue, it would be nice to remove it because of code bloat and blocking other optimizations. 

**Merge basic blocks**
The single block optimization is a lot more powerful than the other ones because it knows more about the code flow and usages of the `alloc_stack`. It would be good to either add another pass to merge basic blocks before this or add a simple merge to this pass (like `mergeBasicBlocks` in `MandatoryInlining`).

**Ownership loads**
I've decided that for this patch I'm not supporting `load [take]` at all and I'm not supporting `load [copy]` in alloc to phi. I've disabled optimizing non-trivial ownership loads in the alloc to phi optimization because it's pretty difficult to optimize those correctly (though, I would like to support them eventually). One of the examples I ran into while I was trying to support ownership loads in alloc to phi is the following:
```
// Before the pass
sil hidden [ossa] @test_no_leak_consuming_post_dom : $@convention(method) (@guaranteed String, @owned Builtin.BridgeObject) -> () {
bb0(%1 : @guaranteed $String, %2: @owned $Builtin.BridgeObject):
  %8 = alloc_stack $String
  %9 = copy_value %1 : $String
  store %9 to [init] %8 : $*String
  %18 = alloc_stack $Builtin.BridgeObject
  store %2 to [init] %18 : $*Builtin.BridgeObject
  br bb1
bb1:
  cond_br undef, bb2, bb3
bb2:
  %35 = load [copy] %8 : $*String
  destroy_value %35 : $String
  %42 = copy_value %1 : $String
  store %42 to [assign] %8 : $*String
  br bb1
bb3:
  destroy_addr %18 : $*Builtin.BridgeObject
  destroy_addr %8 : $*String
  dealloc_stack %18 : $*Builtin.BridgeObject
  dealloc_stack %8 : $*String
  %55 = tuple ()
  return %55 : $()
}
// After the pass
sil hidden [ossa] @test_no_leak_consuming_post_dom : $@convention(method) (@guaranteed String, @owned Builtin.BridgeObject) -> () {
// %0                                             // users: %7, %2
// %1                                             // user: %3
bb0(%0 : @guaranteed $String, %1 : @owned $Builtin.BridgeObject):
  %2 = copy_value %0 : $String                    // user: %4
  destroy_value %1 : $Builtin.BridgeObject        // id: %3
  br bb1(%2 : $String)                            // id: %4
// %5                                             // user: %9
bb1(%5 : @owned $String):                         // Preds: bb0 bb2
  cond_br undef, bb2, bb3                         // id: %6
bb2:                                              // Preds: bb1
  %7 = copy_value %0 : $String                    // user: %8
  br bb1(%7 : $String)                            // id: %8
bb3:                                              // Preds: bb1
  %9 = destructure_struct %5 : $String            // user: %10
  %10 = destructure_struct %9 : $_StringGuts      // user: %11
  (%11, %12) = destructure_struct %10 : $_StringObject // user: %13
  destroy_value %12 : $Builtin.BridgeObject       // id: %13
  %14 = tuple ()                                  // user: %15
  return %14 : $()                                // id: %15
} // end sil function 'test_no_leak_consuming_post_dom'
```
As you can see, `%5` is leaked. I would need to add a destroy somewhere in `bb2` but, I'm not sure what logic I should use to figure out when an extra destroy is needed and where it should go. It would likely be expensive. 

Supporting `load [take]` should be easier, especially in single blocks. I think I can essentially apply the same logic as `load [copy]` where I simply remove destroys. Balancing the needed destroys may be more difficult, though. 

---
I'm also going to try to add this higher up in the pass pipeline so that it can actually see some `ossa`.

zoecarver committed:
Add mem2reg before ownership elimination

zoecarver commented:
@swift-ci please benchmark.

zoecarver committed:
Add extra arcopts pass after mem2reg

zoecarver commented:
@swift-ci please benchmark.

zoecarver commented on lib/SILOptimizer/PassManager/PassPipeline.cpp:
```diff
@@ -400,6 +400,9 @@ static void addPerfEarlyModulePassPipeline(SILPassPipelinePlan &P) {
   P.addSemanticARCOpts();
 
+  P.addMem2Reg();
+  P.addSemanticARCOpts();
```
This semantics ARC opt seems to really help clean up after `mem2reg`. I'll look into moving some of those optimizations into mem2reg so that we don't have to duplicate this pass.

zoecarver commented:
@swift-ci Please benchmark

zoecarver committed:
Reset changes to the pass pipeline

zoecarver commented:
@atrick @gottesmm As you've probably noticed, I've been cleaning up and simplifying all my ownership PRs so that you can (hopefully) review them more easily. Here's the dilemma with this change, PHI arguments are proving to be extremely difficult with ownership. That's the reason both for the regressions and the added complexity in this PR. 

However, I don't think that the PHI arguments are much better than a dominance-based implementation. Also as expressed above, this pass makes assumptions that are hard to have in OSSA and I suspect could be broken with enough effort in non-ownership mode (post dominance leaks, and using `undef` as a placeholder seem particularly fragile).

This pass could be dramatically simplified and more "natively" support OSSA if re-written. I would like to discuss re-writting this pass as a combiner. My load promotion (#30463) and store elimination (#30549) PRs do much of the mem2reg optimizations and could be easily moved into a new pass. I know you both have a lot on your plate right now and I'm worried that reviewing such a large change will be difficult. What do you think? Like the idea? Think it's feasible?

zoecarver commented:
I'm going to close this PR for now. Updating this pass in the right way will take time and energy from not just myself but others. And that time and energy is better spent on other things right now. When the time comes to update this pass, please feel free to use anything from this patch. In this PR I've documented the things that need to be updated and the challenges that come with various approaches. Also, there are quite a few tests that mostly came from various errors when compiling the stdlib. Please take those too :)
